Feature/imports

Thomas-McKanna opened the pull request:
Curious to see what you all think about this pull request. First and foremost, this pull request removes all relative imports.

Second, how tasks are imported has changed. Every new task added was resulting in another `from tasks.x.y.z import Z` being added to `dronecontroller.py`. I moved the contents of the `task` folder into its `__init__.py`, which greatly simplified how tasks can be imported. It is similar to how dronekit does it (https://github.com/dronekit/dronekit-python/blob/master/dronekit/__init__.py).

Thomas-McKanna committed:
Removed relative imports from all files.

Thomas-McKanna committed:
Merge branch 'develop' into feature/imports

Thomas-McKanna committed:
Merge branch 'develop' into feature/imports

Thomas-McKanna committed:
Tasks easy to import.

christopher-o-toole commented on flight/drone/drone.py:
```diff
@@ -133,8 +133,7 @@ def arm(self, mode=c.Modes.GUIDED.value):
         while not self.armed:
-            if self.is_armable:
-                self.armed = True
+            self.armed = True
```
Is there a particular reason why this is preferred over checking if the drone is armable?

Thomas-McKanna replied:
Yeah, so it turns out that drone.vehicle.is_armable is just a bugged value (see https://github.com/dronekit/dronekit-python/issues/695 and https://discuss.ardupilot.org/t/is-armable-parameter-doesnt-change-dronekit-python/21574), or its just not being used right. It always returns False, and so it prevents takeoff.

I had John submit a pull request to add checking for is_armable, but we never actually tested it (not good).

christopher-o-toole commented on flight/tasks/__init__.py:
```diff
@@ -0,0 +1,207 @@
+import abc
```
docstring should be added to all brand new files from here on out

christopher-o-toole commented on flight/tasks/__init__.py:
```diff
@@ -0,0 +1,207 @@
+from flight import flightconfig as f
+from flight.drone.exceptions import EmergencyLandException
+
+KP = 1
```
we need a comment for each of the constants in this group indicating what they are/what their purpose is.

christopher-o-toole commented on flight/tasks/__init__.py:
```diff
@@ -0,0 +1,207 @@
+        """
+        super(Exit, self).__init__(drone)
+
+    def perform(self):
```
Missing function docstring.

christopher-o-toole commented on flight/tasks/__init__.py:
```diff
@@ -0,0 +1,207 @@
+        self._pid_alt = PID(KP, KI, KP, setpoint=altitude)
+        self._count = duration * (1.0/c.DELAY_INTERVAL)
+
+    def perform(self):
```
Missing function docstring

christopher-o-toole commented on flight/tasks/__init__.py:
```diff
@@ -0,0 +1,207 @@
+        super(Land, self).__init__(drone)
+        self._land_mode = VehicleMode(c.Modes.LAND.value)
+
+    def perform(self):
```
Missing function docstring.

christopher-o-toole commented on flight/tasks/__init__.py:
```diff
@@ -0,0 +1,207 @@
+        self._vy = velocities[1]
+        self._vz = velocities[2]
+
+    def perform(self):
```
Missing function docstring.

christopher-o-toole commented on flight/tasks/__init__.py:
```diff
@@ -0,0 +1,207 @@
+        """
+        super(Takeoff, self).__init__(drone)
+        self._target_alt = altitude
+    def perform(self):
```
Missing function docstring.

christopher-o-toole reviewed (changes requested):
I do think that getting rid of relative imports is ideal maintainability wise, however it may be somewhat more inconvenient to maintain many tasks which are all thrown into the same file. Perhaps consider having a tasks module folder in which that module's `__init__.py` imports all officially released tasks into the relevant namespace.

jozgd committed:
Better task importing

jozgd committed:
moved the user_command.py and removed the is_armable

ClayMav commented on flight/drone/drone_controller.py:
```diff
@@ -8,16 +8,14 @@
-from ... import flightconfig as f
+
+from flight import constants as c
+import flight.tasks
```
```suggestion
from flight.tasks import Hover, Takeoff, LinearMovement, Land, Exit
```

ClayMav commented on user_command.py:
```diff
@@ -55,7 +55,7 @@ def matches(self, *args):
         if len(args) != len(self._expected_order):
             raise TypeError('Expected {} arguments, got {}.'.format(
-                len(args),  len(self._expected_order)))
+                len(args),  len(self._expected_order)))#might need to flip these
```
Yeah this should be flipped fam

ClayMav reviewed (changes requested):
Hit a fix fam, then we good dawg

Thomas-McKanna commented:
I did the requested changes. In addition I have done the following:

- Added docstrings to _all_ files which I have written the majority of
- Fixed bug with giving a hover command in user_command.py
- Renamed user_command.py to test_cli.py

Code changes have been tested and no problems were encountered.

Thomas-McKanna committed:
Added file docstrings. Small bug fixes.

christopher-o-toole reviewed (dismissed):
Much better.

ClayMav reviewed (changes requested):
I am happy but Codacy is not, just fix some of those Codacy complaints and I will approve

Thomas-McKanna committed:
Codacy-recommended changes.

Thomas-McKanna committed:
Merge branch 'develop' into feature/imports
